ext/datasource/refreshable_file/file_datasource.go

```diff
@@ -0,0 +1,114 @@
+package refreshable_file
+
+import (
+	"github.com/alibaba/sentinel-golang/ext/datasource"
+	"github.com/alibaba/sentinel-golang/logging"
+	"github.com/alibaba/sentinel-golang/util"
+	"github.com/fsnotify/fsnotify"
+	"github.com/pkg/errors"
+	"io/ioutil"
+	"os"
+	"time"
+)
+
+var (
+	logger = logging.GetDefaultLogger()
+)
+
+type RefreshableFileDataSource struct {
+	datasource.Base
+	sourceFilePath string
+	isInitialized  util.AtomicBool
+	closeChan      chan struct{}
+	watcher        *fsnotify.Watcher
+}
+
+func NewFileDataSource(sourceFilePath string, handlers ...datasource.PropertyHandler) (*RefreshableFileDataSource, error) {
+	var ds = &RefreshableFileDataSource{
+		sourceFilePath: sourceFilePath,
+		closeChan:      make(chan struct{}),
+	}
+	for _, h := range handlers {
+		ds.AddPropertyHandler(h)
```

Comment:
Do we need to check nil here?

Reply:
I think it's not necessary, if nil, AddPropertyHandler will not append nil.

Reply:
But we can explicitly check nil in advance here

Reply:
Okay for me if `AddPropertyHandler` has handled the case.